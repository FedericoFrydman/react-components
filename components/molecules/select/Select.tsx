import { forwardRef, ReactNode, Ref } from 'react';
import * as RadixSelect from '@radix-ui/react-select';
import cn from 'classnames';

import { Icon } from '../../../components/atoms/icon/Icon';
import { InputValidationStatus } from '../input/Input';

import './select.scss';

export const SelectGroup = (props: RadixSelect.SelectGroupProps) => <RadixSelect.Group {...props} />;
export const SelectLabel = (props: RadixSelect.SelectLabelProps) => (
  <RadixSelect.Label className="lsux-select-item--label" {...props} />
);
export const SelectSeparator = (props: RadixSelect.SelectSeparatorProps) => (
  <RadixSelect.Separator className="lsux-select-item--separator" {...props} />
);

export interface SelectItemProps extends RadixSelect.SelectItemProps {
  classNames?: cn.Argument;
}

export const SelectItem = forwardRef<HTMLDivElement, SelectItemProps>(
  ({ children, classNames, ...props }, forwardedRef) => {
    let textValue = props.textValue || '';
    if (typeof children === 'string' && children.length > 0) {
      textValue = children;
    }

    return (
      <RadixSelect.Item
        className={cn('lsux-select-item', classNames)}
        disabled={props.disabled}
        value={props.value}
        textValue={textValue}
        ref={forwardedRef}
        {...props}
      >
        <RadixSelect.ItemIndicator className="lsux-select-item--indicator">
          <Icon name="interface_check" />
        </RadixSelect.ItemIndicator>
        <RadixSelect.ItemText>{textValue}</RadixSelect.ItemText>
      </RadixSelect.Item>
    );
  },
);
SelectItem.displayName = 'SelectItem';

export interface SelectProps extends RadixSelect.SelectProps {
  /**
   * Value of the select component.
   */
  value?: string;
  /**
   * Triggers error | warning | success styles on the component. Important for accessibility.
   * @type {InputValidationStatus}
   * @default undefined
   */
  status?: InputValidationStatus;
  /**
   * Placeholder text to display.
   * @default 'Select an option'
   */
  placeholder?: string;
  /**
   * Items to render in the select component.
   * @description This expects the `SelectItem` component.
   */
  children?: ReactNode;
  /**
   * Stretch the select component to fill the width of the parent container.
   * @default false
   */
  stretch?: boolean;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLButtonElement>;
}

const SelectComponent = (
  {
    children,
    value,
    placeholder = 'Select an option',
    status,
    stretch = false,
    classNames = [],
    ...props
  }: SelectProps,
  ref: SelectProps['ref'],
) => {
  const classSelectTrigger = 'lsux-select__trigger';
  const cnSelectTrigger = cn({
    [`${classSelectTrigger}`]: true,
    [`${classSelectTrigger}--${status}`]: status,
    [`${classSelectTrigger}--stretch`]: stretch,
  });

  return (
    <RadixSelect.Root onValueChange={props.onValueChange} value={value} disabled={props.disabled} {...props}>
      <RadixSelect.Trigger
        ref={ref}
        aria-label={props.name}
        className={cn(cnSelectTrigger, classNames)}
        placeholder={placeholder}
      >
        <div className="lsux-select__trigger__row">
          <RadixSelect.Value className="lsux-select__trigger__value" placeholder={placeholder} />
          <RadixSelect.Icon className="lsux-select__trigger__icon">
            <Icon name="arrow_caret_down_md" />
          </RadixSelect.Icon>
        </div>
      </RadixSelect.Trigger>
      <RadixSelect.Portal>
        <RadixSelect.Content className="lsux-select__content">
          <RadixSelect.ScrollUpButton className="lsux-select__content__button">
            <Icon name="arrow_caret_up_md" />
          </RadixSelect.ScrollUpButton>
          <RadixSelect.Viewport className="lsux-select__viewport">{children}</RadixSelect.Viewport>
          <RadixSelect.ScrollDownButton className="lsux-select__content__button">
            <Icon name="arrow_caret_down_md" />
          </RadixSelect.ScrollDownButton>
          <RadixSelect.Arrow />
        </RadixSelect.Content>
      </RadixSelect.Portal>
    </RadixSelect.Root>
  );
};

export const Select = forwardRef<HTMLButtonElement, SelectProps>(SelectComponent);
