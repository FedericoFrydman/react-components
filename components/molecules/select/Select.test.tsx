import { shallow, ShallowWrapper } from 'enzyme';
import { Select, SelectItem, SelectItemProps } from './Select';

const options: SelectItemProps[] = [
  {
    textValue: 'United States',
    value: 'US',
  },
  {
    textValue: 'Germany',
    value: 'DE',
  },
  {
    textValue: 'France',
    value: 'FR',
  },
];

describe('Select Component', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Select placeholder="Select a country">
        {options.map((c) => (
          <SelectItem key={c.value} {...c} />
        ))}
      </Select>,
    );
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
