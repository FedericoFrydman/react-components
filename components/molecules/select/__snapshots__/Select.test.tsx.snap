// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component should render the component 1`] = `
<Select>
  <SelectTrigger
    className="lsux-select__trigger"
    placeholder="Select a country"
  >
    <div
      className="lsux-select__trigger__row"
    >
      <SelectValue
        className="lsux-select__trigger__value"
        placeholder="Select a country"
      />
      <SelectIcon
        className="lsux-select__trigger__icon"
      >
        <ForwardRef(IconComponent)
          name="arrow_caret_down_md"
        />
      </SelectIcon>
    </div>
  </SelectTrigger>
  <SelectPortal>
    <SelectContent
      className="lsux-select__content"
    >
      <SelectScrollUpButton
        className="lsux-select__content__button"
      >
        <ForwardRef(IconComponent)
          name="arrow_caret_up_md"
        />
      </SelectScrollUpButton>
      <SelectViewport
        className="lsux-select__viewport"
      >
        <SelectItem
          key="US"
          textValue="United States"
          value="US"
        />
        <SelectItem
          key="DE"
          textValue="Germany"
          value="DE"
        />
        <SelectItem
          key="FR"
          textValue="France"
          value="FR"
        />
      </SelectViewport>
      <SelectScrollDownButton
        className="lsux-select__content__button"
      >
        <ForwardRef(IconComponent)
          name="arrow_caret_down_md"
        />
      </SelectScrollDownButton>
      <SelectArrow />
    </SelectContent>
  </SelectPortal>
</Select>
`;
